README.md
<p align="center"><img src="https://raw.githubusercontent.com/emsk/redmine-now/master/resources/redmine-now-logo.png" alt="Redmine Now"></p>

<p align="center">A desktop app to know what's happening now on your Redmine</p>

<p align="center">
  <a href="https://travis-ci.org/emsk/redmine-now"><img src="https://travis-ci.org/emsk/redmine-now.svg" alt="Build Status"></a>
  <a href="https://gemnasium.com/github.com/emsk/redmine-now"><img src="https://gemnasium.com/badges/github.com/emsk/redmine-now.svg" alt="Dependency Status"></a>
  <a href="https://github.com/sindresorhus/xo"><img src="https://img.shields.io/badge/code_style-XO-5ed9c7.svg" alt="XO code style"></a>
  <a href="https://github.com/emsk/redmine-now/blob/master/LICENSE"><img src="https://img.shields.io/badge/license-MIT-blue.svg" alt="License"></a>
</p>

---

![Screenshot Main](examples/redmine-now.png?raw=true)

![Screenshot Settings](examples/redmine-now-settings.png?raw=true)

### Dark Mode

You can toggle dark mode with <kbd>Cmd</kbd> <kbd>Shift</kbd> <kbd>D</kbd> / <kbd>Ctrl</kbd> <kbd>Shift</kbd> <kbd>D</kbd>.

![Screenshot Dark Mode Main](examples/redmine-now-dark.png?raw=true)

![Screenshot Dark Mode Settings](examples/redmine-now-dark-settings.png?raw=true)

## Installation

Installers for macOS and Windows can be found on the [releases](../../releases) page.

## Development

Built with [Electron](http://electron.atom.io).

## Related

* [Redmine Notifier](https://github.com/emsk/redmine-notifier) - Updated issues checker for Redmine

## License

[MIT](LICENSE)


CHANGELOG.md
# Change Log

## 0.4.0 (2017-02-15)

* Change installer for Windows from Squirrel to NSIS
* Change app ID
* Assign cancel action to escape key on the settings window
* Change scrollbar style on Windows

## 0.3.0 (2017-01-30)

* Extract settings into child window
* Add dark mode
* Show overlay button when `Redmine URL` or `API key` is empty
* Add `Window` and `Redmine Now Website` to app menu
* Change hover background color of issue elements
* Hide spin button in text fields
* Customize title bar on macOS
* Update Electron to 1.4.15

## 0.2.0 (2017-01-24)

* Automatically update issue list 
* Redesign appearance
* Show `assigned_to`, `updated_on`, and total issue count
* Add `About Redmine Now` menu
* Restore previous window state
* Do not fetch when `Redmine URL` or `API key` is empty
* Simplify internal process of showing issues
* Fix order of issues
* Fix bug that more than 25 issues are not fetched at once
* Update Electron to 1.4.13

## 0.1.0 (2017-01-01)

* When `Fetch` button is clicked, fetch and show issues
* When `Show/Hide settings` button is clicked, show or hide settings
* Add app menu
* Add logo and icons


test/test.js
'use strict';

import test from 'ava';
import {Application} from 'spectron';

test.beforeEach(async t => {
  t.context.app = new Application({
    path: `${__dirname}/../dist/mac/Redmine Now.app/Contents/MacOS/Redmine Now`
  });

  await t.context.app.start();
});

test.afterEach.always(async t => {
  await t.context.app.stop();
});

test(async t => {
  const app = t.context.app;
  await app.client.waitUntilWindowLoaded();

  const client = app.client;
  t.is(await client.getWindowCount(), 1);
  t.regex(await client.getUrl(), /^file:\/\/.+\/index.html$/);
  t.is(await client.getTitle(), 'Redmine Now');

  const win = app.browserWindow;
  const {width, height} = await win.getBounds();
  t.true(width > 0);
  t.true(height > 0);

  t.deepEqual(await win.getMinimumSize(), [300, 200]);
  t.true(await win.isVisible());
  t.true(await win.isResizable());
  t.true(await win.isFocused());
  t.false(await win.isMaximized());
  t.false(await win.isMinimized());
  t.false(await win.isFullScreen());
  t.true(await win.isMovable());
  t.true(await win.isMaximizable());
  t.true(await win.isMinimizable());
  t.true(await win.isFullScreenable());
  t.true(await win.isClosable());
  t.false(await win.isAlwaysOnTop());
  t.false(await win.isKiosk());
  t.false(await win.isDocumentEdited());
  t.false(await win.isMenuBarAutoHide());
  t.true(await win.isMenuBarVisible());
  t.false(await win.isVisibleOnAllWorkspaces());
  t.false(await win.isDevToolsOpened());
  t.false(await win.isDevToolsFocused());
});



app/main.js
'use strict';

(() => {
  const electron = require('electron');
  const app = electron.app;
  const BrowserWindow = electron.BrowserWindow;
  const Config = require('electron-config');
  const config = new Config();

  const isMac = process.platform === 'darwin';

  let win = null;

  app.on('window-all-closed', () => {
    app.quit();
  });

  // Avoid the slow performance issue when renderer window is hidden
  app.commandLine.appendSwitch('disable-renderer-backgrounding');

  app.on('ready', () => {
    let winOptions = {
      title: 'Redmine Now',
      show: false,
      width: 850,
      height: 670,
      minWidth: 300,
      minHeight: 200,
      titleBarStyle: isMac ? 'hidden' : 'default'
    };
    Object.assign(winOptions, config.get('winBounds'));
    win = new BrowserWindow(winOptions);

    win.loadURL(`file://${__dirname}/index.html`);

    win.once('ready-to-show', () => {
      win.show();
    });

    win.on('close', () => {
      config.set('winBounds', win.getBounds());
    });

    win.on('closed', () => {
      win = null;
    });
  });
})();



app/menu.js
'use strict';

const electron = require('electron');

const remote = electron.remote;
const app = remote.app;
const dialog = remote.dialog;
const shell = remote.shell;
const Menu = remote.Menu;

const isMac = process.platform === 'darwin';

const appName = app.getName();
const appWebsite = 'https://github.com/emsk/redmine-now';
const appCopyright = 'Copyright (c) 2016-2017 emsk';
const appIconFilePath = `${__dirname}/images/redmine-now-icon.png`;

const sendAction = action => {
  remote.getCurrentWindow().webContents.send(action);
};

let appMenuItems = [
  {
    label: 'Edit',
    submenu: [
      {role: 'undo'},
      {role: 'redo'},
      {type: 'separator'},
      {role: 'cut'},
      {role: 'copy'},
      {role: 'paste'},
      {role: 'selectall'}
    ]
  },
  {
    role: 'window',
    submenu: [
      {role: 'minimize'},
      {role: 'zoom'},
      {type: 'separator'},
      {role: 'front'},
      {role: 'togglefullscreen'}
    ]
  }
];

const preferencesMenuItem = {
  label: 'Preferences...',
  accelerator: 'CmdOrCtrl+,',
  click: () => {
    sendAction('open-settings-window');
  }
};

const toggleDarkModeMenuItem = {
  label: 'Toggle Dark Mode',
  accelerator: 'CmdOrCtrl+Shift+D',
  click: () => {
    sendAction('toggle-dark-mode');
  }
};

const websiteMenuItem = {
  label: `${appName} Website`,
  click: () => {
    shell.openExternal(appWebsite);
  }
};

if (isMac) {
  appMenuItems.unshift({
    label: app.getName(),
    submenu: [
      {role: 'about'},
      {type: 'separator'},
      preferencesMenuItem,
      {type: 'separator'},
      toggleDarkModeMenuItem,
      {type: 'separator'},
      {role: 'quit'}
    ]
  });

  appMenuItems.push({
    role: 'help',
    submenu: [
      websiteMenuItem
    ]
  });
} else {
  appMenuItems.unshift({
    label: 'File',
    submenu: [
      preferencesMenuItem,
      {type: 'separator'},
      toggleDarkModeMenuItem,
      {type: 'separator'},
      {role: 'quit'}
    ]
  });

  appMenuItems.push({
    role: 'help',
    submenu: [
      websiteMenuItem,
      {type: 'separator'},
      {
        label: `About ${appName}`,
        click: () => {
          dialog.showMessageBox({
            title: `About ${appName}`,
            message: `${appName} ${app.getVersion()}`,
            detail: appCopyright,
            icon: appIconFilePath,
            buttons: []
          });
        }
      }
    ]
  });
}

module.exports = Menu.buildFromTemplate(appMenuItems);



app/util.js
'use strict';

const isMac = process.platform === 'darwin';

module.exports.hideTitleBar = () => {
  if (isMac) {
    return false;
  }

  const titleBarElement = document.getElementById('title-bar');
  if (titleBarElement !== null) {
    titleBarElement.classList.add('windows');
  }

  const openSettingsButtonElement = document.getElementById('open-settings-button');
  if (openSettingsButtonElement !== null) {
    openSettingsButtonElement.classList.add('windows');
  }

  const mainElement = document.getElementById('main');
  if (mainElement !== null) {
    mainElement.classList.add('windows');
  }

  return true;
};

module.exports.formatDate = (date, todayTime = null) => {
  const year = date.getFullYear();
  const month = `0${date.getMonth() + 1}`.slice(-2);
  const day = `0${date.getDate()}`.slice(-2);
  const hour = `0${date.getHours()}`.slice(-2);
  const minute = `0${date.getMinutes()}`.slice(-2);
  const dateTime = new Date(year, date.getMonth(), day).getTime();

  if (todayTime === dateTime) {
    return `${hour}:${minute}`;
  }
  return `${year}-${month}-${day} ${hour}:${minute}`;
};



app/settings.js
'use strict';

(() => {
  const util = require('./util');
  const electron = require('electron');
  const ipcRenderer = electron.ipcRenderer;
  const remote = electron.remote;

  const defaultUpdateIntervalSec = 600;
  const baseTimeDaysAgo = 14;

  class Setting {
    constructor(startupTime) {
      this._startupTime = startupTime;
    }

    initEventListener() {
      ipcRenderer.on('toggle-dark-mode', (event, isDarkMode) => {
        document.body.classList.toggle('dark', isDarkMode);
      });

      document.addEventListener('keydown', event => {
        if (event.key === 'Escape') {
          remote.getCurrentWindow().close();
        }
      });

      document.getElementById('save-settings-button').addEventListener('click', () => {
        this.updateSettings()
          .updateLastExecutionTimeWithBaseTime();
        remote.getCurrentWindow().getParentWindow().webContents.send('save-settings');
        remote.getCurrentWindow().close();
      });

      document.getElementById('cancel-settings-button').addEventListener('click', () => {
        remote.getCurrentWindow().close();
      });

      return this;
    }

    displayDefaultSettings() {
      document.getElementById('default-update-interval').innerHTML = defaultUpdateIntervalSec;

      const select = document.getElementById('base-time');
      const option = document.createElement('option');
      option.value = -1;
      option.innerText = util.formatDate(this._startupTime);
      select.appendChild(option);

      for (let i = 0; i <= baseTimeDaysAgo; i++) {
        const date = new Date();
        date.setDate(date.getDate() - i);
        const option = document.createElement('option');
        option.value = i;
        option.innerText = util.formatDate(new Date(date.getFullYear(), date.getMonth(), date.getDate()));
        select.appendChild(option);
      }

      return this;
    }

    displaySettings() {
      document.getElementById('url').value = localStorage.getItem('url');
      document.getElementById('api-key').value = localStorage.getItem('apiKey');
      document.getElementById('project-id').value = localStorage.getItem('projectId');
      document.getElementById('update-interval').value = localStorage.getItem('updateInterval');

      const baseTimeValue = localStorage.getItem('baseTimeValue');
      if (baseTimeValue !== null) {
        document.getElementById('base-time').value = baseTimeValue;
      }

      return this;
    }

    updateSettings() {
      localStorage.setItem('url', document.getElementById('url').value);
      localStorage.setItem('apiKey', document.getElementById('api-key').value);
      localStorage.setItem('projectId', document.getElementById('project-id').value);
      localStorage.setItem('updateInterval', document.getElementById('update-interval').value);

      const selectedOption = this.getSelectedBaseTime();
      const date = new Date(selectedOption.text);
      const baseTime = date.toISOString().replace(/\.\d+Z$/, 'Z');
      localStorage.setItem('baseTime', baseTime);
      localStorage.setItem('baseTimeValue', selectedOption.value);

      return this;
    }

    updateLastExecutionTimeWithBaseTime() {
      const selectedOption = this.getSelectedBaseTime();
      const date = new Date(selectedOption.text);
      const baseTime = date.toISOString().replace(/\.\d+Z$/, 'Z');

      localStorage.setItem('lastExecutionTime', baseTime);

      return this;
    }

    getSelectedBaseTime() {
      const select = document.getElementById('base-time');
      return select.options[select.selectedIndex];
    }
  }

  ipcRenderer.on('load-settings-window', (event, startupTime) => {
    const appStartupTime = new Date(startupTime);

    const setting = new Setting(appStartupTime);
    setting.initEventListener()
      .displayDefaultSettings()
      .displaySettings();
  });

  document.addEventListener('DOMContentLoaded', () => {
    const isDarkMode = JSON.parse(localStorage.getItem('isDarkMode'));
    document.body.classList.toggle('dark', isDarkMode);

    util.hideTitleBar();
  });
})();



app/index.js
'use strict';

(() => {
  const appMenu = require('./menu');
  const util = require('./util');
  const electron = require('electron');
  const ipcRenderer = electron.ipcRenderer;
  const remote = electron.remote;
  const BrowserWindow = remote.BrowserWindow;
  const shell = remote.shell;
  const Menu = remote.Menu;

  const isMac = process.platform === 'darwin';

  const defaultUpdateIntervalSec = 600;

  let isDarkMode = JSON.parse(localStorage.getItem('isDarkMode'));

  class RedmineNow {
    constructor() {
      this._settingsWindow = null;
      this._settings = {};
      this._startupTime = null;
      this._needsUpdateStatus = true;
      this._issueStatuses = [];
      this._timer = null;
      this._issues = [];
    }

    initMenu() {
      Menu.setApplicationMenu(appMenu);
      return this;
    }

    initEventListener() {
      ipcRenderer.on('toggle-dark-mode', () => {
        this.toggleDarkMode();
      });

      ipcRenderer.on('open-settings-window', () => {
        this.openSettingsWindow();
      });

      ipcRenderer.on('save-settings', () => {
        this._needsUpdateStatus = true;
        this.readStoredSettings()
          .overlay()
          .initFetch();
      });

      document.getElementById('open-settings-button').addEventListener('click', () => {
        this.openSettingsWindow();
      });

      remote.getCurrentWindow().on('close', () => {
        this.removeBaseTime();
      });

      return this;
    }

    openSettingsWindow() {
      if (this._settingsWindow !== null) {
        return this;
      }

      this._settingsWindow = new BrowserWindow({
        title: 'Settings',
        show: false,
        resizable: false,
        maximizable: false,
        fullscreenable: false,
        width: isMac ? 540 : 555,
        height: isMac ? 240 : 265,
        parent: isMac ? null : remote.getCurrentWindow(),
        titleBarStyle: isMac ? 'hidden' : 'default'
      });

      if (!isMac) {
        this._settingsWindow.setMenuBarVisibility(false);
      }

      this._settingsWindow.loadURL(`file://${__dirname}/settings.html`);

      this._settingsWindow.once('ready-to-show', () => {
        if (isMac) {
          this._settingsWindow.setParentWindow(remote.getCurrentWindow());
        }
        this._settingsWindow.show();
      });

      this._settingsWindow.on('closed', () => {
        this._settingsWindow = null;
      });

      this._settingsWindow.webContents.on('did-finish-load', () => {
        this._settingsWindow.webContents.send('load-settings-window', this._startupTime);
      });

      return this;
    }

    toggleDarkMode() {
      isDarkMode = document.body.classList.toggle('dark');
      localStorage.setItem('isDarkMode', isDarkMode);

      if (this._settingsWindow !== null) {
        this._settingsWindow.webContents.send('toggle-dark-mode', isDarkMode);
      }
    }

    initStartupTime() {
      this._startupTime = new Date();
      return this;
    }

    fetchIssueStatus() {
      if (!this._needsUpdateStatus) {
        return this;
      }

      if (!this.validateSettings()) {
        return this;
      }

      const xhr = new XMLHttpRequest();

      xhr.onreadystatechange = () => {
        if (xhr.readyState === 4) {
          this.handleResponseFetchIssueStatus(xhr.status, xhr.responseText);
        }
      };

      this.updateLastExecutionTimeWithBaseTime();

      xhr.open('GET', `${this._settings.url}/issue_statuses.json`);
      xhr.setRequestHeader('X-Redmine-API-Key', this._settings.apiKey);
      xhr.send();

      return this;
    }

    handleResponseFetchIssueStatus(status, responseText) {
      if (status === 200) {
        const headers = document.getElementById('headers');
        const container = document.getElementById('container');
        this._issueStatuses = JSON.parse(responseText).issue_statuses;

        this.clear();

        this._issueStatuses.forEach(issueStatus => {
          const header = document.createElement('div');
          header.id = `header-column-status-${issueStatus.id}`;
          header.className = 'header-column';
          header.innerText = issueStatus.name;
          headers.appendChild(header);

          const column = document.createElement('div');
          column.id = `column-status-${issueStatus.id}`;
          column.className = isMac ? 'column' : 'column windows';
          container.appendChild(column);
        });

        this._needsUpdateStatus = false;
      }

      return this;
    }

    clear() {
      this._issues = [];
      document.getElementById('headers').innerHTML = '';
      document.getElementById('container').innerHTML = '';
    }

    initFetch() {
      const timer = () => {
        this.fetch(1);
        clearInterval(this._timer);
        this._timer = setInterval(timer, this.getUpdateIntervalMsec());
      };

      clearInterval(this._timer);
      this._timer = setInterval(timer, this.getUpdateIntervalMsec());

      return this;
    }

    getUpdateIntervalMsec() {
      return 1000 * (this._settings.updateInterval || defaultUpdateIntervalSec);
    }

    fetch(page) {
      this.fetchIssueStatus();

      if (!this.validateSettings()) {
        return this;
      }

      const xhr = new XMLHttpRequest();

      xhr.onreadystatechange = () => {
        if (xhr.readyState === 4) {
          this.handleResponseFetch(xhr.status, xhr.responseText, page);
        }
      };

      xhr.open('GET', `${this._settings.url}/issues.json${this.getRequestParams(page)}`);
      xhr.setRequestHeader('X-Redmine-API-Key', this._settings.apiKey);
      xhr.send();

      return this;
    }

    handleResponseFetch(status, responseText, page) {
      if (status === 200) {
        const response = JSON.parse(responseText);

        this.keepIssues(response.issues);

        if (this.calcRemainingRequestCount(response, page) > 0) {
          this.fetch(page + 1);
        } else {
          this.sortIssuesByUpdatedOn()
            .showIssues()
            .showTotalIssue()
            .updateLastExecutionTime();
        }
      }

      return this;
    }

    getRequestParams(page) {
      const lastExecutionTime = localStorage.getItem('lastExecutionTime');
      const params = [
        `updated_on=%3E%3D${lastExecutionTime}`,
        'status_id=*',
        'sort=updated_on:asc',
        'limit=100',
        `page=${page}`
      ];

      const projectId = this._settings.projectId;
      if (projectId !== null && projectId !== '') {
        params.unshift(`project_id=${projectId}`);
      }

      return `?${params.join('&')}`;
    }

    calcRemainingRequestCount(response, page) {
      return Math.floor(response.total_count / response.limit) + 1 - page;
    }

    keepIssues(issues) {
      issues.forEach(issue => {
        this.removeIssue(issue.id);
        this._issues.push(issue);
      });

      return this;
    }

    removeIssue(issueId) {
      this._issues = this._issues.filter(issue => {
        return issue.id !== issueId;
      });

      return this;
    }

    sortIssuesByUpdatedOn() {
      this._issues.sort((a, b) => {
        const dateA = new Date(a.updated_on);
        const dateB = new Date(b.updated_on);
        return dateA - dateB || a.id - b.id;
      });

      return this;
    }

    showIssues() {
      const issueCount = this._issues.length;

      if (issueCount === 0) {
        return this;
      }

      const now = new Date();
      const todayTime = new Date(now.getFullYear(), now.getMonth(), now.getDate()).getTime();

      this._issues.forEach(issue => {
        const issueElementId = `issue-${issue.id}`;
        this.removeIssueElement(issueElementId);
        const issueElement = this.createIssueElement(issueElementId, issue, todayTime, this._settings.url);
        const column = document.getElementById(`column-status-${issue.status.id}`);
        column.insertBefore(issueElement, column.firstChild);
      });

      return this;
    }

    removeIssueElement(issueElementId) {
      const issueElement = document.getElementById(issueElementId);
      if (issueElement) {
        issueElement.parentNode.removeChild(issueElement);
      }

      return this;
    }

    createIssueElement(issueElementId, issue, todayTime, url) {
      const issueElement = document.createElement('div');
      issueElement.id = issueElementId;
      issueElement.className = 'issue';

      const issueIdElement = document.createElement('div');
      issueIdElement.innerText = `#${issue.id}`;
      issueIdElement.className = 'issue-id';
      issueElement.appendChild(issueIdElement);

      const subjectElement = document.createElement('div');
      subjectElement.innerText = issue.subject;
      subjectElement.className = 'subject';
      issueElement.appendChild(subjectElement);

      const assignedTo = issue.assigned_to;
      if (assignedTo !== undefined) {
        const assignedToElement = document.createElement('div');
        assignedToElement.innerText = assignedTo.name;
        assignedToElement.className = 'assigned-to';
        issueElement.appendChild(assignedToElement);
      }

      const updatedOnElement = document.createElement('div');
      updatedOnElement.innerText = util.formatDate(new Date(issue.updated_on), todayTime);
      updatedOnElement.className = 'updated-on';
      issueElement.appendChild(updatedOnElement);

      issueElement.addEventListener('click', () => {
        shell.openExternal(`${url}/issues/${issue.id}`);
      });

      return issueElement;
    }

    showTotalIssue() {
      const issueCount = this._issues.length;

      document.getElementById('total-issue-count').innerText = issueCount;

      const issueUnit = (issueCount === 0 || issueCount === 1) ? 'issue' : 'issues';
      document.getElementById('total-issue-unit').innerText = issueUnit;

      return this;
    }

    updateLastExecutionTime(date = new Date()) {
      const lastExecutionTime = date.toISOString().replace(/\.\d+Z$/, 'Z');
      localStorage.setItem('lastExecutionTime', lastExecutionTime);

      return this;
    }

    updateLastExecutionTimeWithBaseTime() {
      if (this._settings.baseTime !== null) {
        this.updateLastExecutionTime(new Date(this._settings.baseTime));
      }

      return this;
    }

    readStoredSettings() {
      this._settings = {
        url: localStorage.getItem('url'),
        apiKey: localStorage.getItem('apiKey'),
        projectId: localStorage.getItem('projectId'),
        updateInterval: localStorage.getItem('updateInterval'),
        baseTime: localStorage.getItem('baseTime')
      };

      return this;
    }

    validateSettings() {
      const url = this._settings.url;
      const apiKey = this._settings.apiKey;

      if (url === null || url === '' || apiKey === null || apiKey === '') {
        return false;
      }

      return true;
    }

    removeBaseTime() {
      localStorage.removeItem('baseTime');
      localStorage.removeItem('baseTimeValue');

      return this;
    }

    overlay() {
      const url = this._settings.url;
      const apiKey = this._settings.apiKey;
      const overlayElement = document.getElementById('no-setting');

      if (url === null || url === '' || apiKey === null || apiKey === '') {
        overlayElement.classList.add('overlay');
      } else {
        overlayElement.classList.remove('overlay');
      }

      return this;
    }
  }

  document.addEventListener('DOMContentLoaded', () => {
    document.body.classList.toggle('dark', isDarkMode);

    util.hideTitleBar();
  });

  window.addEventListener('load', () => {
    const redmineNow = new RedmineNow();
    redmineNow.initMenu()
      .initEventListener()
      .initStartupTime()
      .readStoredSettings()
      .overlay()
      .fetchIssueStatus()
      .updateLastExecutionTime(redmineNow._startupTime)
      .initFetch();
  });
})();

